<a href="https://colab.research.google.com/github/marianaarboleda/pc-para-todas/blob/master/visualizacion_widgets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Visualización de datos
## Ahora con código 🥰

Hasta el momento hemos hablado de por qué es importante visualizar, algunas fuentes de datos y cómo podemos crear malas representaciones de los datos.


Ahora vamos a explorar un nuevo set de datos con la ayuda de este cuaderno interactivo de python.

Primero, descarga los datos: https://drive.google.com/file/d/10BqIEzrdGBH_mGYRuQ8iMoEPVt1qlMcU/view?usp=sharing

Fuente original: https://www.datos.gov.co/Transporte/accidentes-calle-30-2015-2019/sefb-a755


Las celdas se ejecutan utilizando el botón de "play" o simplemente con shift+enter.

Siempre empezamos por importar los paquetes necesarios. En este caso: pandas, plotly y los widgets

In [ ]:
import pandas as pd
import plotly.express as px

import ipywidgets as widgets

La celda que sigue, hace parte de la magia de plotly y nos permite que las gráficas sean interactivas

In [ ]:
config_chart = {
    'scrollZoom': True, 'displaylogo': False, 'responsive': True,
    'editable': True,
    'toImageButtonOptions': {
        'format': 'png',  # one of png, svg, jpeg, webp
        'filename': 'custom_image',
        'height': None,
        'width': None,
        'scale': 6  # Multiply title/legend/axis/canvas sizes by this factor
    }
}

Leemos los datos

In [ ]:
df = pd.read_csv('/content/Accidentalidad_en_Barranquilla.csv')
df.shape

(42178, 11)

la función head imprime


In [ ]:
df.head(5)

FECHA_ACCIDENTE AÑO_ACCIDENTE MES_ACCIDENTE DIA_ACCIDENTE  \
0  01/01/2015 12:00:00 AM         2,015         enero           Jue   
1  01/01/2015 12:00:00 AM         2,015         enero           Jue   
2  01/01/2015 12:00:00 AM         2,015         enero           Jue   
3  01/01/2015 12:00:00 AM         2,015         enero           Jue   
4  01/01/2015 12:00:00 AM         2,015         enero           Jue   

  HORA_ACCIDENTE GRAVEDAD_ACCIDENTE CLASE_ACCIDENTE SITIO_EXACTO_ACCIDENTE  \
0    03:30:00:PM        Con heridos          Choque            CL 45 CR 19   
1    04:20:00:AM         Solo daños          Choque         CRA 15 CLLE 21   
2    04:40:00:PM        Con heridos          Choque         CRA 14 CLLE 35   
3    04:50:00:PM        Con heridos       Atropello          CRA 6 CLLE 90   
4    06:00:00:AM         Solo daños          Choque          CRA 6 CLLE 92   

   CANT_HERIDOS_EN _SITIO_ACCIDENTE  CANT_MUERTOS_EN _SITIO_ACCIDENTE  \
0                               2.0                               NaN   
1                               NaN                               NaN   
2                               2.0                               NaN   
3                               1.0                               NaN   
4                               NaN                               NaN   

   CANTIDAD_ACCIDENTES  
0                    1  
1                    1  
2                    1  
3                    1  
4                    1

In [ ]:
## Guardamos los nombres de las columnas
col_list = df.columns.tolist()

Abajo creamos nuestros selectores de columnas.

In [ ]:
drop_list = [None] + col_list

y_axis = "CANTIDAD_ACCIDENTES"

x_axis = widgets.Dropdown(options=drop_list)
color = widgets.Dropdown(options=drop_list)
agg_func = widgets.Dropdown(options=['count','mean','sum'])


display("Eje x: ", x_axis)
display("Dividir por color: ", color)
display("Agrupador: ", agg_func)



'Eje x: '

Dropdown(options=(None, 'FECHA_ACCIDENTE', 'AÑO_ACCIDENTE', 'MES_ACCIDENTE', 'DIA_ACCIDENTE', 'HORA_ACCIDENTE'…

'Dividir por color: '

Dropdown(options=(None, 'FECHA_ACCIDENTE', 'AÑO_ACCIDENTE', 'MES_ACCIDENTE', 'DIA_ACCIDENTE', 'HORA_ACCIDENTE'…

'Agrupador: '

Dropdown(options=('count', 'mean', 'sum'), value='count')

Prueba a graficar AÑO_ACCIDENTE en el eje X. Luego divide según la gravedad del accidente. Recuerda ejecutar la celda de abajo, cada vez que realices un cambio

In [ ]:
index_cols = list(set([x_axis.value, color.value]).intersection(set(col_list)))


pivot = df.pivot_table(index=index_cols, aggfunc=agg_func.value, values=y_axis).reset_index()

fig = px.bar(pivot, x=x_axis.value, y=y_axis, color=color.value)
fig.show(config=config_chart)

### Chao selectores

Las interfaces gráficas son útiles, pero a veces nos limitan.

In [ ]:
col_list

['FECHA_ACCIDENTE',
 'AÑO_ACCIDENTE',
 'MES_ACCIDENTE',
 'DIA_ACCIDENTE',
 'HORA_ACCIDENTE',
 'GRAVEDAD_ACCIDENTE',
 'CLASE_ACCIDENTE',
 'SITIO_EXACTO_ACCIDENTE',
 'CANT_HERIDOS_EN _SITIO_ACCIDENTE',
 'CANT_MUERTOS_EN _SITIO_ACCIDENTE',
 'CANTIDAD_ACCIDENTES']

In [ ]:
y_axis = "MES_ACCIDENTE"
x_axis =  'CLASE_ACCIDENTE'
color = 'GRAVEDAD_ACCIDENTE'
facet_row = None
facet_col = None

agg_func = 'first' #"sum", "nunique", "mean"

index_cols = list(set([x_axis, color, facet_row, facet_col]).intersection(set(col_list)))

pivot = df.pivot_table(index=index_cols, aggfunc=agg_func, values=y_axis).reset_index()

fig = px.bar(pivot, x=x_axis, y=y_axis, color=color,
             facet_row=facet_row, facet_col=facet_col,
             color_discrete_sequence=px.colors.qualitative.Pastel,
             barmode='group',
             text=y_axis
             )

fig.update_traces(textposition='outside', texttemplate='%{text}')
fig.show(config=config_chart)

Programar es 92% aprender a leer documentación. Elige la paleta de colores que más te guste y cámbiala

https://plotly.com/python/discrete-color/